### ISSUE TITLE
Memory usage in parallel minRMSD clustering

### ISSUE THREAD

**Issue by jchodera** [markovmodel/PyEMMA#776]
I'm testing the recent #770 parallel clustering additions on a dataset with ~1M snapshots using 32 processes, and find it violates the 96GB memory limit I give it after ~12 hours.

Is it expected that memory usage should slowly grow during clustering, or are there provisions made to keep memory usage minimal when chunking is in use?

If the expectation is that memory usage should remain minimal, I wonder if there is an unintentional memory leak. We frequently encounter cases where relying on Python garbage collection leads to slow growth in memory usage, where we have to explicitly `del` objects after we're done with them, even if all references fall out of scope.

**Comment by marscher** [comment]
I'd guess we have introduced a memleak. @clonker: we free the dists 
pointer for every thread, right?

**Comment by marscher** [comment]
Yes. There is a memleak, please stand by.

**Comment by marscher** [comment]
This went on devel.

**Comment by marscher** [comment]
Valgrind now reports no leaks in Pyemma's code (only im OpenMM initialization, couple of bytes, nothing to worry about).

**Comment by jchodera** [comment]
Thanks!  So I should just be able to update `master` and reinstall?

**Comment by jchodera** [comment]
Do you guys have a `pyemma-dev` conda package built from `master`?

**Comment by marscher** [comment]
Precisely.

**Comment by jchodera** [comment]
Testing now. Thanks!

**Comment by franknoe** [comment]
I don't think we have a pyemma-dev build, or do we @marscher?

It might be useful to have that though - wouldn't that even be a 
replacement for release candidates?

Am 19/04/16 um 18:02 schrieb John Chodera:

> Testing now. Thanks!
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/776#issuecomment-211995429

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by marscher** [comment]
I think @jchodera is talking about nightly builds via http://nightli.es/ 

A nightly build is in my opinion a unstable version, which changes very frequently in comparison to a release candidate.

**Comment by jchodera** [comment]
We add a small script to upload the tested package to anaconda cloud at the end of the `.travis.yml`, [like this](https://github.com/choderalab/yank/blob/master/.travis.yml#L48-L50). These packages are named `packagename-dev`, and are used by other dev recipes to make sure that you can quickly catch changes in upstream packages that break your own package. These are indeed unstable builds, but can be very useful in both catching upstream changes that break things and in deploying the latest dev branch to test on some other hardware where manually installing dependencies and building the package may be a pain.

**Comment by marscher** [comment]
I really like this idea! Is there any effort going on to standardize 
this for omnia?

**Comment by jchodera** [comment]
It's mostly standardized. See the [devtools](https://github.com/omnia-md/devtools) guidelines and recommended scripts.

**Comment by marscher** [comment]
Alright. I've enabled it. Thank you very much!

**Comment by franknoe** [comment]
Cool, thanks. How is this installed? Just `conda install pyemma-dev`, if the omnia channel is known?

**Comment by marscher** [comment]
Yes - that should do it.

**Comment by jchodera** [comment]
I can confirm that `pyemma-dev` is being pushed to [anaconda cloud](https://anaconda.org/omnia/pyemma-dev/files)---thanks!

I can also confirm the memory leak is gone.  It still takes ~40 hours for minRMSD clustering 1M configurations using equitemporally-selected generators on 32 threads, but it seems to work robustly now!  Many thanks.

**Comment by marscher** [comment]
Thanks for the confirmation! And how long did it take before the parallel assignment code, eg. with one thread?

**Comment by franknoe** [comment]
yeah, that's one reason why we usually work with TICA+kmeans. Also, 
time-uniform clustering has performed extremely poor in the tests I have 
made, but there's no systematic analysis of these things as far as a know.

Am 21/04/16 um 17:08 schrieb John Chodera:

> I can confirm that |pyemma-dev| is being pushed to anaconda cloud 
> https://anaconda.org/omnia/pyemma-dev/files---thanks!
> 
> I can also confirm the memory leak is gone. It still takes ~40 hours 
> for minRMSD clustering 1M configurations using equitemporally-selected 
> generators on 32 threads, but it seems to work robustly now! Many thanks.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/776#issuecomment-212963719

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by jchodera** [comment]
Looks like ~29h with 1 thread and ~132h with 32 threads.

**Comment by franknoe** [comment]
I hope the other way around

Am 23/04/16 um 16:56 schrieb John Chodera:

> Looks like ~29h with 1 thread and ~132h with 32 threads.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/776#issuecomment-213753223

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by jchodera** [comment]
Ha!  Yes, the other way around.  Sorry!
